iree/base/api.c

```diff
@@ -0,0 +1,875 @@
+    char* buffer, iree_host_size_t* out_buffer_length);
+
+// Header for optional status payloads.
+// Each status may have zero or more payloads associated with it that can later
+// be used to produce more detailed logging or programmatically query
+// information about an error.
+struct iree_status_payload_s {
+  // Next payload in the status payload linked list.
+  struct iree_status_payload_s* next;
+  // Payload type identifier used for programmatic access to payloads. May be
+  // IREE_STATUS_PAYLOAD_TYPE_OPAQUE if the payload cannot be accessed directly.
+  iree_status_payload_type_t type;
+  // Allocator used for the payload and associated resources.
+  iree_allocator_t allocator;
+  // String formatter callback used to write the payload into a string buffer.
+  // If not present then the payload will be mentioned but not dumped when the
+  // status is logged.
+  iree_status_payload_formatter_t formatter;
+};
+
+// A string message (IREE_STATUS_PAYLOAD_TYPE_MESSAGE).
+typedef struct {
+  iree_status_payload_t header;
+  // String data reference. May point to an address immediately following this
+  // struct (if copied) or a constant string reference in rodata.
+  iree_string_view_t message;
+} iree_status_payload_message_t;
+
+// Allocated storage for an iree_status_t.
+// Only statuses that have either source information or payloads will have
+// storage allocated for them.
+typedef struct {
+  // Optional doubly-linked list of payloads associated with the status.
+  // Head = first added, tail = last added.
+  iree_status_payload_t* payload_head;
+  iree_status_payload_t* payload_tail;
+
+#if (IREE_STATUS_FEATURES & IREE_STATUS_FEATURE_SOURCE_LOCATION) != 0
+  // __FILE__ of the originating status allocation.
+  const char* file;
+  // __LINE__ of the originating status allocation.
+  uint32_t line;
+#endif  // has IREE_STATUS_FEATURE_SOURCE_LOCATION
+
+#if (IREE_STATUS_FEATURES & IREE_STATUS_FEATURE_ANNOTATIONS) != 0
+  // Optional message that is allocated either as a constant string in rodata or
+  // present as a suffix on the storage.
+  iree_string_view_t message;
+#endif  // has IREE_STATUS_FEATURE_ANNOTATIONS
+} iree_status_storage_t;
+
+// Appends a payload to the storage doubly-linked list.
+static iree_status_t iree_status_append_payload(
+    iree_status_t status, iree_status_storage_t* storage,
+    iree_status_payload_t* payload) {
+  if (!storage->payload_tail) {
+    storage->payload_head = payload;
+  } else {
+    storage->payload_tail->next = payload;
+  }
+  storage->payload_tail = payload;
+  return status;
+}
+
+// Formats an iree_status_payload_message_t to the given output |buffer|.
+// |out_buffer_length| will be set to the number of characters written excluding
+// NUL. If |buffer| is omitted then |out_buffer_length| will be set to the
+// total number of characters in |buffer_capacity| required to contain the
+// entire message.
+static bool IREE_API_CALL iree_status_payload_message_formatter(
```

Comment:
This is just one formatter of many - other custom formatters may fail (to indicate they have nothing to format or otherwise), such as a stack trace that cannot resolve symbols or a buffer that no longer has access to the data. That said, emitting "<error>" or "" seems fine there, so making void.